### ISSUE TITLE
QR & Image printing very slow

### ISSUE THREAD

**Issue by FransvanHoogstraten** [python-escpos/python-escpos#251]
I have:
- [x] searched open and closed issues for duplicates

Hey guys, 

I have an Epson TM88-III hooked up to a raspberry pi through serial connection (baudrate 9600).

Everything works fine, but the printing of images & QR codes (which are basically images in non-native mode) is very slow with me. Also, I can hear the printer printing the image line by line, with quite some time in between, which I hope is not expected behaviour.

At first I thought this might have to do with time needed for data transfer, but even a 300byte image (which should take <0,5 second to transfer @ 9600 baud) gets printed 'line by line' rather slowly. 

I have made a short video how a QR code currently prints with me:

https://dl.dropboxusercontent.com/u/4651651/2017-08-15%2016.05.26.mov

Is this expected behaviour? Is there any way to make QRs/images print faster/smoothly? 

I am planning on implementing this in a production environment, and I know for sure the local employees will go mad within a day if every receipt is slowed down by the QR printing :)

(PS1: Barcodes are printing quickly & smootly, as they are produced on the printer itself. I have tried setting the flag native=True with QR generation, which does not seem to work.)
 
(PS2: I also found this thread which describes the same problem, just with a different library. Any thoughts on this? https://github.com/Simonefardella/escposprinter/issues/4)

**Comment by FransvanHoogstraten** [comment]
So, after some thorough testing I did get a bit further. 

It seems that printing QR/images goes alright if I make sure the image is a simple bitmap (no compression or greyscales), and no bigger than approximately 100*100 pixels.

Then if I use high_density_vertical/horizontal=False the image comes out at reasonable size, at OK speed (@9600 baud)

**Comment by patkan** [comment]
Hi,
which version of the library are you using? We have sped up the processing in the past, but you probably already have that version.
Depending on your processing power it might take a short moment. To time this better you could use the dummy-printer to print into memory. Then you definitely can rule out the transfer.
I haven't tested the library recently on a system in the league of Raspis so I am not sure what is normal, but we had a thread once where iirc @mike42 timed the execution of the single parts of the image-printing

**Comment by FransvanHoogstraten** [comment]
Hi Patrick, 

Thanks for the reply!

So I've tried the preprocessing with the dummy printer, but this actually doesn't help anything. Therefore I'm pretty sure the cause was that I was EITHER using images which were too big (causing the data-transfer to be the bottleneck @9600 baud) OR I used images with too high compression (certain PNG types) causing the delay. 

It seems now the problem has ceased now I switched to pure 1-bit bitmap images (no greyscale!) which are no larger than 100*100 pixels.

Thanks though!
